### PR TITLE
[ADD] Subsampling Dataset

### PR DESCRIPTION
**Pull request by ravinkohli**
subsamples the dataset if it takes more memory than passed with memory_allocation.
## Types of changes

- [x] New feature (non-breaking change which adds functionality)

Note that a Pull Request should only contain one of refactoring, new features or documentation changes.
Please separate these changes and send us individual PRs for each.
For more information on how to create a good pull request, please refer to [The anatomy of a perfect pull request](https://medium.com/@hugooodias/the-anatomy-of-a-perfect-pull-request-567382bb6067).

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

## Description

This PR adds the ability to subsample dataset in case it takes more memory than allowed by the user. It works by doing a stratifying split using `memory_allocation/ size_of(dataset)` where `memory_allocation` is specified by the user. This PR also moves the dataset_compression to `tabular_validator.py` as it has access to both `X` and `y`. It also fixes the documentation for `dataset_compression`

## Motivation and Context

It allows us to run autopytorch for larger datasets.

## How has this been tested?

I have modified the tests for precision reduction to also test for `subsampling`.

### CODE REVIEW

File: test/test_data/test_validation.py [hunk 1]
```diff
@@ -137,3 +138,50 @@ def test_validation_unsupported():
...
+        'sparse_csr_nan',
+        'sparse_dia_nonan',
+        'sparse_dia_nan',
+        'sparse_dok_nonan',
+        'sparse_dok_nan',
+        'openml_40981',  # Australian
+    ),
+    indirect=True
+)
+def test_featurevalidator_reduce_precision(input_data_featuretest):
```
Comment (ravinkohli) [inline test/test_data/test_validation.py:166]:
```suggestion
def test_featurevalidator_dataset_compression(input_data_featuretest):
```

File: autoPyTorch/data/tabular_validator.py [hunk 2]
```diff
@@ -46,10 +62,59 @@ def __init__(
...
+        Args:
+            X (DatasetCompressionInputType):
+                features of dataset
+            y (SupportedTargetTypes):
+                targets of dataset
+        Returns:
+            DatasetCompressionInputType:
+                Compressed dataset.
+        """
+        is_dataframe = hasattr(X, 'iloc')
```
Comment (nabenabe0928) [inline autoPyTorch/data/tabular_validator.py:93]: ispandas

File: autoPyTorch/pipeline/components/training/trainer/__init__.py [hunk 3]
```diff
@@ -257,6 +257,14 @@ def _fit(self, X: Dict[str, Any], y: Any = None, **kwargs: Any) -> 'TrainerChoic
...
             metrics.extend(get_metrics(dataset_properties=X['dataset_properties'], names=[X['optimize_metric']]))
         additional_losses = X['additional_losses'] if 'additional_losses' in X else None
+
+        # Ensure that the split is not missing any class.
+        labels = X['y_train'][X['backend'].load_datamanager().splits[X['split_id']][0]]
+        if STRING_TO_TASK_TYPES[X['dataset_properties']['task_type']] in CLASSIFICATION_TASKS:
+            unique_labels = len(np.unique(labels))
+            if unique_labels < X['dataset_properties']['output_shape']:
+                raise ValueError(f"Expected number of unique labels {unique_labels} in train split: {X['split_id']}"
+                                 f" to be = num_classes {X['dataset_properties']['output_shape']}.")
```
Comment (nabenabe0928) [inline autoPyTorch/pipeline/components/training/trainer/__init__.py:267]: Could you separate this process as a method `self._get_train_label`?

File: autoPyTorch/data/tabular_feature_validator.py [hunk 4]
```diff
@@ -100,15 +93,6 @@ class TabularFeatureValidator(BaseFeatureValidator):
...
             List of indices of categorical columns
     """
-    def __init__(
-        self,
-        logger: Optional[Union[PicklableClientLogger, Logger]] = None,
-        dataset_compression: Optional[Mapping[str, Any]] = None,
-    ) -> None:
-        self._dataset_compression = dataset_compression
-        self._reduced_dtype: Optional[DatasetDTypeContainerType] = None
-        super().__init__(logger)
```
Comment (nabenabe0928) [inline autoPyTorch/data/tabular_feature_validator.py:110]: Can we keep `__init__` as we need it for dtypes provided by users.?

**Review by nabenabe0928 — CHANGES_REQUESTED** [review]
Hi thanks for the PR, I checked files except tests, tabular_validator, utils.

File: autoPyTorch/data/tabular_feature_validator.py (hunk 4 above)
Comment (ravinkohli) [reply autoPyTorch/data/tabular_feature_validator.py:110]: okay I'll add it back

File: autoPyTorch/data/tabular_feature_validator.py [hunk 5]
```diff
@@ -103,10 +98,7 @@ class TabularFeatureValidator(BaseFeatureValidator):
     def __init__(
         self,
         logger: Optional[Union[PicklableClientLogger, Logger]] = None,
-        dataset_compression: Optional[Mapping[str, Any]] = None,
     ) -> None:
```
Comment (nabenabe0928) [inline autoPyTorch/data/tabular_feature_validator.py:101]:
```suggestion
    ):
```

File: autoPyTorch/pipeline/components/training/trainer/__init__.py [hunk 6]
```diff
@@ -381,6 +384,20 @@ def _fit(self, X: Dict[str, Any], y: Any = None, **kwargs: Any) -> 'TrainerChoic
 
         return self
 
+    def _get_train_label(self, X: Dict[str, Any]) -> List[int]:
+        """
+        Verifies and validates the labels from train split.
+        """
+        # Ensure that the split is not missing any class.
```
Comment (nabenabe0928) [inline autoPyTorch/pipeline/components/training/trainer/__init__.py:391]: What if we could not include some classes in the training split?
I think it can happen when we have few labels for some classes.

File: autoPyTorch/data/tabular_validator.py [hunk 7]
```diff
@@ -27,16 +38,22 @@ class TabularInputValidator(BaseInputValidator):
...
+            documentation for `BaseTask.get_dataset`.
     """
     def __init__(
         self,
         is_classification: bool = False,
         logger_port: Optional[int] = None,
-        dataset_compression: Optional[Mapping[str, Any]] = None,
+        dataset_compression: Optional[DatasetCompressionSpec] = None,
+        seed: int = 42,
     ) -> None:
```
Comment (nabenabe0928) [inline autoPyTorch/data/tabular_validator.py:51]:
```suggestion
    ):
```

**Review by nabenabe0928 — COMMENTED** [review]
Hi thanks for the fixes, I will check utils.py from now on.

File: autoPyTorch/pipeline/components/training/trainer/__init__.py (hunk 6 above)
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/training/trainer/__init__.py:391]: I'd say then either the user should not use a weighted strategy or make a stratified split.

Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/training/trainer/__init__.py:391]: Then we could raise a different kind of error message?

File: autoPyTorch/data/utils.py [hunk 8]
```diff
@@ -283,8 +463,11 @@ def megabytes(arr: DatasetCompressionInputType) -> float:
 
 def reduce_dataset_size_if_too_large(
     X: DatasetCompressionInputType,
-    memory_allocation: int,
-    methods: List[str] = ['precision'],
+    memory_allocation: Union[int, float],
+    is_classification: bool,
+    random_state: Union[int, np.random.RandomState],
```
Comment (nabenabe0928) [inline autoPyTorch/data/utils.py:468]: What should we do for `random_state` in the future?
Should we stick to `np.random.RandomState` or include `int`?
Now everything is `np.random.RandomState`, so I think it is easier to fix it later if we now stick to `np.random.RandomState`

**Review by nabenabe0928 — COMMENTED** [review]
Checked utils.py after Line 466

File: autoPyTorch/data/utils.py [hunk 9]
```diff
@@ -266,6 +368,84 @@ def reduce_precision(
...
+            idxs = X.index[left_idxs]
+            X = X.loc[idxs]
+        else:
+            X = X[left_idxs]
+
+        if isinstance(y, pd.DataFrame) or isinstance(y, pd.Series):
+            idxs = y.index[left_idxs]
+            y = y.loc[idxs]
+        else:
+            y = y[left_idxs]
```
Comment (nabenabe0928) [inline autoPyTorch/data/utils.py:431]: Could you make it as it makes this function easier to track.
```python
def _subsample_by_indices(
    X: Union[np.ndarray, pd.DataFrame],
    y: Union[np.ndarray, pd.DataFrame, pd.Series],
    indices_to_leave: np.ndarray
) -> Tuple[Union[np.ndarray, pd.DataFrame], Union[np.ndarray, pd.DataFrame, pd.Series]]:

    if ispandas(X):
        indices = X.index[indices_to_leave]
        X = X.loc[indices]
    else:
        X = X[indices_to_leave]

    if ispandas(y):
        indices = y.index[indices_to_leave]
        y = y.loc[indices]
    else:
        y = y[indices_to_leave]
```

**Review by nabenabe0928 — COMMENTED** [review]
I still need to check `CustomStratifiedShuffleSplit`

File: autoPyTorch/data/utils.py [hunk 10]
```diff
@@ -35,10 +39,108 @@
...
     "memory_allocation": 0.1,
-    "methods": ["precision"]
+    "methods": ["precision", "subsample"]
 }
 
 
+class CustomStratifiedShuffleSplit(StratifiedShuffleSplit):
+    """Splitter that deals with classes with too few samples"""
+
+    def _iter_indices(self, X, y, groups=None):  # type: ignore
```
Comment (nabenabe0928) [inline autoPyTorch/data/utils.py:49]: typing?

File: autoPyTorch/data/utils.py (hunk 8 above)
Comment (ravinkohli) [reply autoPyTorch/data/utils.py:468]: When the user can instantiate the class or call the method, we should allow both. I think that's similar to sklearn. Internally we can stick with `np.random.RandomState`. What do you think?

File: autoPyTorch/data/utils.py (hunk 9 above)
Comment (ravinkohli) [reply autoPyTorch/data/utils.py:431]: definitely.

File: autoPyTorch/data/utils.py [hunk 11]
```diff
@@ -35,10 +39,108 @@
...
+
+            for i in range(n_classes):
+                permutation = rng.permutation(class_counts[i])
+                perm_indices_class_i = class_indices[i].take(permutation, mode="clip")
+
+                train.extend(perm_indices_class_i[: n_i[i]])
+                test.extend(perm_indices_class_i[n_i[i]: n_i[i] + t_i[i]])
+
+            train = rng.permutation(train)
+            test = rng.permutation(test)
```
Comment (nabenabe0928) [inline autoPyTorch/data/utils.py:139]: Can we make it another method?

File: autoPyTorch/data/utils.py [hunk 12]
```diff
@@ -35,10 +39,108 @@
...
+
+        classes, y_indices = np.unique(y, return_inverse=True)
+        n_classes = classes.shape[0]
+
+        class_counts = np.bincount(y_indices)
+
+        if n_train < n_classes:
+            raise ValueError(
+                "The train_size = %d should be greater or "
+                "equal to the number of classes = %d" % (n_train, n_classes)
```
Comment (nabenabe0928) [inline autoPyTorch/data/utils.py:72]:
```suggestion
                f"The train_size = {n_train} should be greater or "
                "equal to the number of classes = {n_classes}"
```

File: autoPyTorch/data/utils.py [hunk 13]
```diff
@@ -35,10 +39,108 @@
...
+            )
+        if n_test < n_classes:
+            raise ValueError(
+                "The test_size = %d should be greater or "
+                "equal to the number of classes = %d" % (n_test, n_classes)
```
Comment (nabenabe0928) [inline autoPyTorch/data/utils.py:77]:
```suggestion
                f"The test_size = {n_test} should be greater or "
                f"equal to the number of classes = {n_classes}"
```

**Review by nabenabe0928 — COMMENTED** [review]
It is hard to understand this class...

File: autoPyTorch/data/utils.py [hunk 14]
```diff
@@ -266,6 +368,84 @@ def reduce_precision(
...
+            X = X.loc[idxs]
+        else:
+            X = X[left_idxs]
+
+        if isinstance(y, pd.DataFrame) or isinstance(y, pd.Series):
+            idxs = y.index[left_idxs]
+            y = y.loc[idxs]
+        else:
+            y = y[left_idxs]
+    elif y is None:
```
[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/data/utils.py:432]: no. we dont subsample testing data. This check is for unsupervised learning. As autopytorch datasets support not having labeled data, we can have this section future proof.

File: autoPyTorch/data/utils.py (hunk 8 above)
Comment (nabenabe0928) [reply autoPyTorch/data/utils.py:468]: That makes sense.

File: autoPyTorch/data/utils.py (hunk 11 above)
Comment (ravinkohli) [reply autoPyTorch/data/utils.py:139]: Please check [this](https://github.com/automl/Auto-PyTorch/pull/398#discussion_r820940363) comment

File: autoPyTorch/data/utils.py [hunk 15]
```diff
@@ -511,20 +524,19 @@ def reduce_dataset_size_if_too_large(
     """
 
     for method in methods:
+        if megabytes(X) > memory_allocation:
```
Comment (nabenabe0928) [inline autoPyTorch/data/utils.py:527]: Thanks for addressing my comment, but we should break the loop when we get `megabytes(X) <= memory_allocation`.
By doing it (see [here](https://github.com/automl/Auto-PyTorch/pull/398#discussion_r820704868)), we can reduce the nest.

**Review by nabenabe0928 — COMMENTED** [review]
Thanks for addressing my comments.
I left a comment!

File: test/test_data/test_utils.py [hunk 16]
```diff
@@ -22,23 +36,128 @@
...
+        assert list(dtype(y_sampled)) == list(dtype(y))
+    else:
+        assert dtype(y_sampled) == dtype(y)
+
+    # Function to get the size of an object
+    def size(obj):
+        if isinstance(obj, spmatrix):  # spmatrix doesn't support __len__
+            return obj.shape[0] if obj.shape[0] > 1 else obj.shape[1]
+        else:
+            return len(obj)
```
Comment (nabenabe0928) [inline test/test_data/test_utils.py:143]: This one as well (but probably, if we have this much, we can just make it as a class, named like `SubsampleValidityHelper`)

**Review by nabenabe0928 — CHANGES_REQUESTED** [review]

File: autoPyTorch/data/utils.py [hunk 17]
```diff
@@ -333,7 +521,34 @@ def reduce_dataset_size_if_too_large(
                     f'Dataset too large for allocated memory {memory_allocation}MB, '
                     f'reduced the precision from {dtypes} to {reduced_dtypes}',
                 )
+        elif method == "subsample":
+            # If the dataset is still too big such that we couldn't fit
+            # into the allocated memory, we subsample it so that it does
+            if megabytes(X) > memory_allocation:
```
[the remark this answers is not in the document]
Comment (ravinkohli) [reply autoPyTorch/data/utils.py]: oh sorry, I meant to do this. I'll fix this

File: test/test_data/test_validation.py [hunk 18]
```diff
@@ -137,3 +138,50 @@ def test_validation_unsupported():
...
+    X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(
+        input_data_featuretest, input_data_targets, test_size=0.1, random_state=1)
+    validator = TabularInputValidator(
+        dataset_compression={'memory_allocation': 0.8 * megabytes(X_train), 'methods': ['precision', 'subsample']}
+    )
+    validator.fit(X_train=X_train, y_train=y_train)
+    transformed_X_train, _ = validator.transform(X_train.copy(), y_train.copy())
+
+    assert validator._reduced_dtype is not None
+    assert megabytes(transformed_X_train) < megabytes(X_train)
```
[the remark this answers is not in the document]
Comment (ravinkohli) [reply test/test_data/test_validation.py:178]: so the columns you mean right?

File: test/test_data/test_utils.py [hunk 19]
```diff
@@ -22,23 +36,128 @@
...
+            if isinstance(obj[0], List):
+                return type(obj[0][0])
+            else:
+                return type(obj[0])
+
+        elif isinstance(obj, pd.DataFrame):
+            return obj.dtypes
+
+        else:
+            return obj.dtype
```
[the remark this answers is not in the document]
Comment (ravinkohli) [reply test/test_data/test_utils.py:123]: same as [here](https://github.com/automl/Auto-PyTorch/pull/398#discussion_r821822227)

File: test/test_data/test_utils.py (hunk 16 above)
Comment (ravinkohli) [reply test/test_data/test_utils.py:143]: same as [here](https://github.com/automl/Auto-PyTorch/pull/398#discussion_r821822227)

File: test/test_data/test_validation.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (ravinkohli) [reply test/test_data/test_validation.py:178]: oh so you mean, check if its reduced or not?

File: test/test_data/test_utils.py [hunk 20]
```diff
@@ -22,23 +36,128 @@
...
+        pytest.skip("Can't have pd.Series as y when task is n-dimensional")
+
+    # Convert our data to its given x_type or y_type
+    def convert(arr, objtype):
+        if objtype == np.ndarray:
+            return arr
+        elif objtype == list:
+            return arr.tolist()
+        else:
+            return objtype(arr)
```
[the remark this answers is not in the document]
Comment (nabenabe0928) [reply test/test_data/test_utils.py:95]: Although I understand your feeling because it is just a test, we should probably create a strict rule saying one function one feature.

File: autoPyTorch/data/utils.py (hunk 15 above)
Comment (ravinkohli) [reply autoPyTorch/data/utils.py:527]: yeah I meant to do that. I'll fix this

File: test/test_data/test_utils.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (ravinkohli) [reply test/test_data/test_utils.py:95]: I don't see how that rule will be relevant here. The bigger function here is the test, and we are initialising utility functions needed only for that test here.

[the remark this answers is not in the document]
Comment (nabenabe0928) [reply test/test_data/test_utils.py:95]: Ok, let's stay for now, but let me state my points.
1. I know each function is relevant to the process of this test, but we do not have to know the internal processes
2. If a function or a method becomes longer, it does not fit to the screen without scroll (that is also why flake8 requires us to fit one line to 80 or 120 lines)
3. We can flexibly do it, but when we accept some exceptions, it is hard to define the borderline what situation is allowed or disallowed

Those are my concerns.

[the remark this answers is not in the document]
Comment (ravinkohli) [reply test/test_data/test_utils.py:95]: Although I dont think that we are making an exception as it is pretty clear to us that we dont do something like this in the library. But I agree with your points 1 and 2 and as such, I'll move these three functions outside.

**Comment by nabenabe0928** [comment]
Hi I do not have any other comments, but do you have any idea why `test_pipeline_fit` fails?

**Comment by ravinkohli** [comment]
> Hi I do not have any other comments, but do you have any idea why `test_pipeline_fit` fails?

It's the same reason we discussed, due to scaling the data, we get negative values for the numerical columns and the configuration with `chi2` fails.

**Comment by nabenabe0928** [comment]
Ok, then should we merge it?
If so, you can do it now!
